Provide torch segmentation model via MONAI Deploy

Leengit opened the pull request:
This is a draft pull request at present.  That will change when it is near ready to merge.

ebrahimebrahim commented:
> The MONAI Deploy library currently works for CUDA 11.3 only, which is a dependency that LungAIR otherwise wouldn't have.  (Right?)

Right, currently LungAIR installs torch using light-the-torch, which automatically finds the correct torch wheel given the user's hardware configuration and available CUDA version.

The change [here](https://github.com/KitwareMedical/lungair-desktop-application/pull/41/commits/757028eb507342fa8103c42b656abb7954888893) looks like it forces a wheel url pointing to CUDA 11.3. But the [preinstall hook](https://github.com/KitwareMedical/lungair-desktop-application/blob/f657033d69a76ddaae84049c6a8c1ac01995ed8f/Modules/Scripted/Home/HomeLib/dependency_installer.py#L56) which runs firstt already is going to get a verison that is compatible with the user's system, maybe not 11.3.
So if light-the-torch decides to install say 11.1, then clicking on the "install monai" button would lead to 11.1 wheel getting downloaded and installed, and then the 11.3 wheel also getting downloaded and installed.

My suggestion, if we need to use MONAI Deploy with cuda, is to add an argument to the light-the-torch usage: `--pytorch-computation-backend=cu113`.
This argument would have to be added to the `_executePythonModule` [here](https://github.com/KitwareMedical/lungair-desktop-application/blob/f657033d69a76ddaae84049c6a8c1ac01995ed8f/Modules/Scripted/Home/HomeLib/dependency_installer.py#L56).

> I am thinking to somehow alert the user about this, but to otherwise not install CUDA 11.3 over whatever the user currently has -- probably such installation would be platform dependent, affect other apps, etc.

I agree that it would be good to tell the user that this is happening.
It's unfortunate that using MONAI deploy now limits the CUDA versions and therefore the hardware that LungAIR is compatible with.
Currently we don't even move the xray to GPU before running segmentation.
It's all on CPU, and it's fast enough.
What if we add the argument `--pytorch-computation-backend=cpu` to the `_executePythonModule` line that calls light-the-torch?
Would MONAI depoly work then?
This could remove adding an unnecessary CUDA version dependency to LungAIR.

Leengit commented:
Yes, I think running it on the CPU rather than the GPU might avoid the dependency on CUDA 11.3.  I will give that a try.

> So if light-the-torch decides to install say 11.1, then clicking on the "install monai" button would lead to 11.1 wheel getting downloaded and installed, and then the 11.3 wheel also getting downloaded and installed.

It may be moot if we go with the CPU solution but ... generally I find `pip` to be pretty smart about versions.  We tell `pip` to look first in `https://download.pytorch.org/whl/cu113` but, in the scenario you describe, `pip` may be smart enough to figure out that what's there isn't compatible with 11.1 and thus fall back to the default wheel sources.  I'm not positive -- if it turns out that we need this behavior, I'll be sure to test it first.

Leengit commented:
This implementation is ready for review.  To change the behavior among using the `model0018.pth` file (which was the behavior before this pull request), using the monai-deploy App and `model0018.zip` file, and using the monai-deploy App and `model0018.zip` file within a docker image, change the right-hand side of `Modules/Scripted/Home/HomeLib/segmentation_model.py`, Line 29, which sets `self.model_source` to one of `self.ModelSource.USE_PTH`, `self.ModelSource.USE_LOCAL_ZIP`, or `self.ModelSource.USE_DOCKER_ZIP`.  (Note that the last of these does not work if the CUDA version on your system differs from the CUDA within the docker image, which is 11.3.)

Be sure to click the "Check for MONAI install" the first time you use this LungAIR, because it checks for and installs additional Python packages now.  The LungAIR interface is otherwise unchanged.  Use it as you would normally use it.

ebrahimebrahim reviewed (commented):
Nice, the `USE_LOCAL_ZIP` mode worked for me!
It was slow however, taking a couple of seconds to do one segmentation on my machine, where the old `USE_PTH` seemed instant. I wonder why it's slower? Maybe because the `USE_PTH` approach holds the segmentation model in memory while LungAIR is running, while the monai deploy approach needs to load the model into memory in a new process at each inference. I don't know much about monai deploy; is there a way around this?

The `USE_DOCKER_ZIP` did not work; expand details to see my error:
<details>

```txt
Checking for MAP "lung_air_app:latest" locally
"lung_air_app:latest" not found locally.

Trying to pull from registry...
Error response from daemon: pull access denied for lung_air_app, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
ERROR: Unable to fetch required image.
Execution Aborted
Traceback (most recent call last):
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 171, in onSegmentSelectedClicked
    self.logic.segmentSelected()
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 493, in segmentSelected
    self.xray_collection.segment_selected()
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 439, in segment_selected
    self.selected_xray().add_segmentation()
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 204, in add_segmentation
    self.seg_mask_tensor, model_to_image_matrix = self.seg_model.run_inference(self.get_numpy_array())
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/segmentation_model.py", line 133, in run_inference
    seg_processed = torch.from_numpy(np.asarray(PIL.Image.open(output_mask_path)))
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/python-install/lib/python3.9/site-packages/PIL/Image.py", line 2953, in open
    fp = builtins.open(filename, "rb")
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmporvzus2_/mask.png'
```

</details>
I guess the docker image needs to be made public

Leengit commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -60,20 +86,54 @@ def run_inference(self, img):
+
+      model_to_img_matrix = np.diag(np.array(img.shape)/self.image_size)
+
+      # TODO returning early because post processing causes crash due to ITK python issues
```
Will do.

ebrahimebrahim commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -2,44 +2,70 @@
+      seg_net = model_dict['model']
+      # set dropout and batch normalization layers to evaluation mode before running
+      # inference
+      tmp = seg_net.eval()
```
Hmm are you sure it was outputting statements to the console? I don't think this should print anything. Maybe it only happened because you were in an interactive python session? (you know that feature that interactive sessions often have where they print out the return value of your last expression if you don't assign it to anything)

Leengit replied:
Yes, it was an interactive Python session.  I will remove the assignment.

Leengit commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -60,20 +86,54 @@ def run_inference(self, img):
+        # Note: to create the docker image, use
+        # f"monai-deploy package {deploy_app_path} --tag {docker_image} --model {self.save_zip_path}"
+        process_image_command = f"monai-deploy run {docker_image} {input_dir_path} {output_dir_path}"
+      os.system(process_image_command)
```
`monai-deploy` is on the PATH with `os.system` but not with `slicer.util.launchConsoleProcess`.  Not even `python` is on the path with the latter.  Rather than tracking down what is wrong there, I am thinking to stick with `os.system`.

ebrahimebrahim replied:
Sounds good. I see [here](https://github.com/Slicer/Slicer/blob/38e026f75f78320a2cb579b5fa96f0917ab4f02c/Base/Python/slicer/util.py#L3470-L3498) now that when `launchConsoleProcess` is used to execute a python process, one first needs to find `PythonSlicer`... too much trouble.

Leengit commented:
Yes, `USE_LOCAL_ZIP` is substantially slower than `USE_PTH`.  I don't know that there is a solution for this in the present case.

I believe that the slowness is because the machinery of `USE_LOCAL_ZIP` is expensive compared to performing the inference, which is very fast in the present case.  If the inference itself were substantially slower, the scales might tip the other way.  Ultimately the `monai-deploy` approach could be teamed with `MONAI Label` and/or `girder` where the inference and display are on different computers; with substantial performance improvement if otherwise slow inference is done on a powerful computer.

Leengit commented:
> The `USE_DOCKER_ZIP` did not work

Right you are, apologies.  I didn't upload the image to anywhere that it could be found automatically.  I will fix that.

If you want to locally build the docker image yourself, which is slow the first time, the command is given in a comment in `segmentation_model.py`:
```shell
monai-deploy package {deploy_app.py_path} --tag lung_air_app:latest --model {model0018.zip_path}
```
where you use the actual paths rather than those `{file_path}` strings.

andinet commented:
I tested your PR on a Windows system

Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.1.4

(Full system information below ** )

1) deploy_app is not correctly copied to the binary directory

Slicer-build\lib\LungAIR-4.13\qt-scripted-modules\HomeLib

I had to manually copy it to try your deploy options.

2) The USE_LOCAL_ZIP option doesn't seem to work. There is no error message 

In the "segmentation_model.py", line 130, you are using os.system to run the monai-deploy command 

      os.system(process_image_command)

May be you might consider to use subprocess module to capture the error message better.




PS

(**) 
Microsoft Visual Studio Community 2022
Version 17.1.4
VisualStudio.17.Release/17.1.4+32407.343
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

Visual C++ 2022   00482-90000-00000-AA257
Microsoft Visual C++ 2022

ASP.NET and Web Tools 2019   17.1.358.51495
ASP.NET and Web Tools 2019

Azure App Service Tools v3.0.0   17.1.358.51495
Azure App Service Tools v3.0.0

C# Tools   4.1.0-5.22165.10+e555772db77ca828b02b4bd547c318387f11d01f
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   17.0.21344.1
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   6.1.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

NVIDIA CUDA 11.7 Wizards   11.7
Wizards to create new NVIDIA CUDA projects and source files.

NVIDIA Nsight Visual Studio Edition   2022.2.0.22095
NVIDIA Nsight Visual Studio Edition provides tools for GPGPU and graphics development.  Copyright © NVIDIA 2010 - 2022.
•Direct3D® and DirectX® are registered trademarks of Microsoft Corporation in the United States and/or other countries.
•Microsoft Detours is used under the Professional license (http://research.microsoft.com/en-us/projects/detours/).
•Gardens Point Parser Generator Copyright 2005 Queensland University of Technology (QUT). All rights reserved.
•Icons from Axialis Software used under the licensing terms found here: www.axialis.com
•NLog Copyright © 2004-2006 Jaroslaw Kowalski (jaak@jkowalski.net)
•zlib and libpng used under the zlib/libpnc license (http://opensource.org/licenses/Zlib) 
•Breakpad Copyright ©2006, Google Inc. All rights reserved.
•The OpenGL Extension Wrangler Library
Copyright ©2008-2016, Nigel Stewart (nigels@users.sourceforge.net), Copyright ©2002-2008, Milan Ikits (milan.ikits@ieee.org), Copyright ©2002-2008, Marcelo E. Magallon (mmagallo@debian.org), Copyright ©2002, Lev Povalahev. 
All rights reserved. 
•LIBSSH2 Copyright ©2004-2007 Sara Golemon (sarag@libssh2.org), Copyright ©2005,2006 Mikhail Gusarov (dottedmag@dottedmag.net),Copyright ©2006-2007 The Written Word, Inc.,Copyright ©2007 Eli Fant (elifantu@mail.ru),Copyright ©2009-2014 Daniel Stenberg., Copyright ©2008, 2009 Simon Josefsson.
All rights reserved. 
•Protobuf Copyright ©2014, Google Inc. All rights reserved.
•xxHASH Library Copyright ©2012-2014, Yann Collet. All rights reserved. 
•FMT Copyright ©2012 - 2016, Victor Zverovich 
•Font Awesome Copyright 2018 Fonticons, Inc. 
•ELF Definitions Copyright (c) 2010 Joseph Koshy, All rights reserved.

Warning: This computer program is protected by copyright law and international treaties. Unauthorized reproduction or distribution of this program, or any portion of it, may result in severe civil and criminal penalties, and will be prosecuted to the maximum extent possible under the law.

NVIDIA Nsight Visual Studio Edition - CUDA support   2022.2.0.22095
NVIDIA Nsight Visual Studio Edition - CUDA support provides tools for CUDA development and debugging.

Python - Django support   17.0.21344.1
Provides templates and integration for the Django web framework.

Python - Profiling support   17.0.21344.1
Profiling support for Python projects.

Python - VC Project Support   17.0.21344.1
Provides support for launching C++ projects with Python debugging enabled.

Python with Pylance   17.0.21344.1
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   17.0.1229.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.1.0-5.22165.10+e555772db77ca828b02b4bd547c318387f11d01f
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

ebrahimebrahim commented:
> 1. deploy_app is not correctly copied to the binary directory

@Leengit Regarding this point here are the cmake lines that you'd have to supplement in order to copy `deploy_app` and `model018.zip` during build: https://github.com/KitwareMedical/lungair-desktop-application/blob/f657033d69a76ddaae84049c6a8c1ac01995ed8f/Modules/Scripted/Home/CMakeLists.txt#L6-L28

jcfr commented:
> consider to use subprocess module to capture the error message better.

I suggest to look into `slicer.util.launchConsoleProcess` and `slicer.util.logProcessOutput`.
See https://discourse.slicer.org/t/invoking-docker-from-a-slicer-module/24529/2?u=jcfr

Leengit commented:
I am trying with
```python
proc = slicer.util.launchConsoleProcess(process_image_command)
slicer.util.logProcessOutput(proc)
```
but failing with
`ModuleNotFoundError: No module named '_sysconfigdata__linux_x86_64-linux-gnu'`
and nothing I have tried is getting me past that.

I have tried `slicer.util.launchConsoleProcess()` with simpler commands, which work.  I'll keep trying, but if that's something you've seen before, please let me know.

jcfr commented:
> but failing with ModuleNotFoundError: No module named '_sysconfigdata__linux_x86_64-linux-gnu' and nothing I have tried is getting me past that.

Are you able to ...

* run the following command in the (1) lung air custom application and (2) the latest Slicer release:
  
  ```python
  import shutil
  dockerExecutablePath = shutil.which('docker')
  commandLine = [dockerExecutablePath, "run", "--rm", "hello-world:latest"]
  proc = slicer.util.launchConsoleProcess(commandLine)
  slicer.util.logProcessOutput(proc)
  ```


> with simpler commands, which work

Do you have an example of such command ? I am wondering if the message comes from the process that is being executed.

Leengit commented:
How do I tell what version of Slicer the LungAIR app is using? ` slicer.__version__` doesn't exist.

@jcfr your example use of `slicer.util.logProcessOutput(proc)` does work.  <details><summary>The command I am trying still isn't working:</summary>
```python
[
    "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/bin/monai-deploy",
    "exec",
    "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/deploy_app.py",
    "-m",
    "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Resources/PyTorchModels/LungSegmentation/model0018.zip",
    "-i",
    "/tmp/tmpzbul0w0y",
    "-o",
    "/tmp/tmpjfeh61fy",
]
```
(I am not actually using `~` symbols.  I put those in the above for brevity and anonymity -- I actually use the full path returned by `shutil.which()` in each case.)
</details>

<details><summary>The full output is:</summary>

```
Traceback (most recent call last):
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/bin/monai-deploy", line 5, in <module>
    from monai.deploy.cli.main import main
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/site-packages/monai/__init__.py", line 38, in <module>
    from .utils.module import load_submodules  # noqa: E402
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/site-packages/monai/utils/__init__.py", line 15, in <module>
    from .deprecate_utils import DeprecatedError, deprecated, deprecated_arg
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/site-packages/monai/utils/deprecate_utils.py", line 19, in <module>
    from monai.utils.module import version_leq
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/site-packages/monai/utils/module.py", line 21, in <module>
    from pydoc import locate
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/pydoc.py", line 458, in <module>
    class Doc:
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/pydoc.py", line 488, in Doc
    def getdocloc(self, object, basedir=sysconfig.get_path('stdlib')):
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/sysconfig.py", line 524, in get_path
    return get_paths(scheme, vars, expand)[name]
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/sysconfig.py", line 514, in get_paths
    return _expand_vars(scheme, vars)
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/sysconfig.py", line 175, in _expand_vars
    _extend_dict(vars, get_config_vars())
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/sysconfig.py", line 564, in get_config_vars
    _init_posix(_CONFIG_VARS)
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/lib/python3.9/sysconfig.py", line 430, in _init_posix
    _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
ModuleNotFoundError: No module named '_sysconfigdata__linux_x86_64-linux-gnu'
Traceback (most recent call last):
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 171, in onSegmentSelectedClicked
    self.logic.segmentSelected()
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 493, in segmentSelected
    self.xray_collection.segment_selected()
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 439, in segment_selected
    self.selected_xray().add_segmentation()
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 204, in add_segmentation
    self.seg_mask_tensor, model_to_image_matrix = self.seg_model.run_inference(self.get_numpy_array())
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/segmentation_model.py", line 142, in run_inference
    slicer.util.logProcessOutput(proc)
  File "~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/bin/Python/slicer/util.py", line 3276, in logProcessOutput
    raise CalledProcessError(retcode, proc.args, output=proc.stdout, stderr=proc.stderr)
subprocess.CalledProcessError: Command '['~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/python-install/bin/monai-deploy', 'exec', '~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/deploy_app.py', '-m', '~/git/KitwareMedical/lungair-SuperBuild-Debug-Qt5.15.2/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Resources/PyTorchModels/LungSegmentation/model0018.zip', '-i', '/tmp/tmpgliqf4f2', '-o', '/tmp/tmpzaerqc4r']' returned non-zero exit status 1.
```
</details>

The exact same command works via `os.system(' '.join(args))`.

jcfr commented:
Since the command you are executing requires the Slicer environment, considering passing `useStartupEnvironment=False` when calling `launchConsoleProcess`.

This should allows the python packages installed in the Slicer environment to be found and used when calling `monai-deploy`

Leengit commented:
> considering passing `useStartupEnvironment=False` when calling `launchConsoleProcess`.

That worked, thank you!  I just force-pushed that into the update intended to make Windows work.

ebrahimebrahim commented:
> How do I tell what version of Slicer the LungAIR app is using? ` slicer.__version__` doesn't exist.

A specific Slicer commit is pinned in the following line: https://github.com/KitwareMedical/lungair-desktop-application/blob/f657033d69a76ddaae84049c6a8c1ac01995ed8f/CMakeLists.txt#L16

jcfr commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -117,21 +118,24 @@ def run_inference(self, img):
+                                 "exec", f"{deploy_app_path}",
                                  "-m", f"{self.save_zip_path}",
                                  "-i", f"{input_dir_path}",
                                  "-o", f"{output_dir_path}"]
```
```suggestion
        process_image_command = [monai_deploy_path,
                                 "exec", deploy_app_path,
                                 "-m", self.save_zip_path,
                                 "-i", input_dir_path,
                                 "-o", output_dir_path]
```

Leengit replied:
Can do.  It seemed "proper" to me to explicitly convert from `os.path` to `str`, which are conceptually different, but this suggestion is definitely cleaner.

jcfr commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -117,21 +118,24 @@ def run_inference(self, img):
-      os.system(' '.join(process_image_command))
+        # Note: to *create* the docker image, use
+        # f"{monai_deploy_path} package {deploy_app_path} --tag {docker_image} --model {self.save_zip_path}"
+        process_image_command = [f"{monai_deploy_path}",
+                                 "run", f"{docker_image}",
+                                 f"{input_dir_path}", f"{output_dir_path}"]
```
```suggestion
        process_image_command = [monai_deploy_path,
                                 "run", docker_image,
                                 input_dir_path, output_dir_path]
```

Leengit replied:
Will do.

jcfr commented:
>  slicer.__version__ doesn't exist.

These may be relevant as well, that said I will suggest to check for features instead of conditionally including code based on a specific version.

Note that in the context of a custom app, you will have different values.

```python
>>> slicer.app.repositoryUrl
'git@github.com:Slicer/Slicer.git'
>>> slicer.app.repositoryRevision
'e3eec68'
>>> slicer.app.mainApplicationMajorVersion
5
>>> slicer.app.mainApplicationMinorVersion
1
>>> slicer.app.mainApplicationPatchVersion
0
>>> slicer.app.mainApplicationRepositoryRevision
'e3eec68'
>>> slicer.app.mainApplicationRepositoryUrl
'git@github.com:Slicer/Slicer.git'
>>> slicer.app.mainApplicationName
'Slicer'
```

jcfr commented:
re: `ENH: Copy binaries for Windows`

Consider updating the commit message to more accurately describe the implemented changes.

Leengit commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -60,20 +87,66 @@ def run_inference(self, img):
+        os.system(' '.join(process_image_command))
+      else:
+        print(f"{process_image_command = }")
+        proc = slicer.util.launchConsoleProcess(process_image_command)
```
Done.

jcfr commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -2,44 +2,70 @@
+
+      self.seg_post_process = SegmentationPostProcessing()
+
+    if self.model_source in (self.ModelSource.USE_LOCAL_ZIP, self.ModelSource.USE_DOCKER_ZIP) and not os.path.exists(self.save_zip_path):
```
Generally speaking, I would suggest again adding binary artifact directly into the version history.

Instead, I would suggest to:
- create an orphan branch, associated tag and a release (e.g models)
- upload the zip file as a release assets
- document the SHA256 of the models in the release description
- download the zip file as part of the build step

Independently of that, the convension used to name the model should be documented as well. Why `model0018` and not `model-YYYY-MM-DD` ?

jcfr replied:
Now, if we don't anticipate the model to change ever ... directly integrating the zip file may be sensible.

Leengit replied:
`model0018.zip` is a member of the repository paralleling the existing membership of `model0018.pth`.  I think the ideas for doing this differently are valuable, for both files, but I'd like to defer these changes to a different pull request.

ebrahimebrahim replied:
Ah, this is probably how I should have handled version control for the model `pth` file; good to know, TY

The name `model0018` relates to the model training experiments in https://github.com/ebrahimebrahim/lung-seg-exploration, which is mentioned at the top of `segmentation_model.py`; I wonder if there is a better place to mention it.

Leengit commented:
I have updated the commit message and made code changes for the above comments.  @andinet would you give it a try on Microsoft Windows?

andinet commented:
The deploy_app is not correctly copied to the binary directory. However, USE_LOCAL_ZIP is still not working.


Python 3.9.10 (main, Aug  1 2022, 05:41:18) [MSC v.1931 64 bit (AMD64)] on win32
Warning: Found an unexpected qMRMLPlotWidget named "qMRMLPlotWidget"; there may be UI setup issues.
Traceback (most recent call last):
  File "D:/b/LM-D-D/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 171, in onSegmentSelectedClicked
    self.logic.segmentSelected()
  File "D:/b/LM-D-D/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 493, in segmentSelected
    self.xray_collection.segment_selected()
  File "D:\b\LM-D-D\Slicer-build\lib\LungAIR-4.13\qt-scripted-modules\HomeLib\xray.py", line 439, in segment_selected
    self.selected_xray().add_segmentation()
  File "D:\b\LM-D-D\Slicer-build\lib\LungAIR-4.13\qt-scripted-modules\HomeLib\xray.py", line 204, in add_segmentation
    self.seg_mask_tensor, model_to_image_matrix = self.seg_model.run_inference(self.get_numpy_array())
  File "D:\b\LM-D-D\Slicer-build\lib\LungAIR-4.13\qt-scripted-modules\HomeLib\segmentation_model.py", line 139, in run_inference
    proc = slicer.util.launchConsoleProcess(process_image_command, useStartupEnvironment=False)
  File "D:\b\LM-D-D\Slicer-build\bin\Python\slicer\util.py", line 3237, in launchConsoleProcess
    proc = subprocess.Popen(args, env=startupEnv, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True, startupinfo=info, cwd=cwd)
  File "D:\b\LM-D-D\python-install\Lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "D:\b\LM-D-D\python-install\Lib\subprocess.py", line 1360, in _execute_child
    args = list2cmdline(args)
  File "D:\b\LM-D-D\python-install\Lib\subprocess.py", line 565, in list2cmdline
    for arg in map(os.fsdecode, seq):
  File "D:\b/LM-D-D/python-install/Lib\os.py", line 822, in fsdecode
    filename = fspath(filename)  # Does type-checking of `filename`.
TypeError: expected str, bytes or os.PathLike object, not NoneType

andinet commented:
@Leengit 

Can you also add a UI ( radio button or something like that) to select one of the model source options?  (USE_PTH, USE_LOCAL_ZIP,   or  USE_DOCKER_ZIP ) i.e expose the options to be changed using the UI.  Manually changing the code and restarting Slicer is not convenient.

You might want to consider adding this menu option to the "Advanced" section.

Leengit commented:
@andinet.  On Ubuntu, I have three different CUDA-related packages that I keep in sync.  Can you give me the versions on Windows that you are using for any of (the counterparts of) these that you have?:
```text
cuda=11.3.1-1
libcudnn8-dev=8.2.1.32-1+cuda11.3
libnvidia-common=465 or 515
```

andinet commented:
C:\Users\andinet.enquobahrie>nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Tue_May__3_19:00:59_Pacific_Daylight_Time_2022
Cuda compilation tools, release 11.7, V11.7.64
Build cuda_11.7.r11.7/compiler.31294372_0

C:\Users\andinet.enquobahrie>

On Tue, Aug 9, 2022 at 9:41 AM Lee Newberg ***@***.***> wrote:

> @andinet <https://github.com/andinet>. On Ubuntu, I have three different
> CUDA-related packages that I keep in sync. Can you give me the versions on
> Windows that you are using for any of (the counterparts of) these that you
> have?:
>
> cuda=11.3.1-1
> libcudnn8-dev=8.2.1.32-1+cuda11.3
> libnvidia-common=465 or 515
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KitwareMedical/lungair-desktop-application/pull/41#issuecomment-1209397968>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADNY5N6XNAQGNKNJEJEOITVYJNXRANCNFSM54T2S5XA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Andinet Enquobahrie, Ph.D., MBA
Director of Medical Computing
Kitware, Inc.

http://www.kitware.com
(919) 969-6990 x311 <%28919%29%20969-6990%20x311>

Leengit commented:
The most recent commit implements a Qt ComboBox in the `Advanced` section of the application so that one can switch between `USE_PTH`, `USE_LOCAL_ZIP`, and `USE_DOCKER_ZIP` without changing the source code and relaunching the application.

andinet commented:
The Qt ComboBox looks good. Thanks.

I built your PR on my Unix machine  (deepak-krs-7910 ) and tried to run the segmentation using the LOCAL_ZIP backend option. I got the error below (*)


andinet @ ~/bin/LungAir-d $nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Wed_Jul_22_19:09:09_PDT_2020
Cuda compilation tools, release 11.0, V11.0.221
Build cuda_11.0_bu.TC445_37.28845127_0
andinet @ ~/bin/LungAir-d $


uname -a
Linux deepak-krs-7910 4.15.0-189-generic #200-Ubuntu SMP Wed Jun 22 19:53:37 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux




PS


(*)



 File "/home/andinet/bin/LungAir-d/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 181, in onSegmentSelectedClicked
    self.logic.segmentSelected(self.backendToUse)
  File "/home/andinet/bin/LungAir-d/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 503, in segmentSelected
    self.xray_collection.segment_selected(backend_to_use)
  File "/home/andinet/bin/LungAir-d/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 444, in segment_selected
    self.selected_xray().add_segmentation(backend_to_use)
  File "/home/andinet/bin/LungAir-d/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 209, in add_segmentation
    self.seg_mask_tensor, model_to_image_matrix = self.seg_model['model'].run_inference(self.get_numpy_array())
  File "/home/andinet/bin/LungAir-d/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/segmentation_model.py", line 137, in run_inference
    proc = slicer.util.launchConsoleProcess(process_image_command, useStartupEnvironment=False)
  File "/home/andinet/bin/LungAir-d/Slicer-build/bin/Python/slicer/util.py", line 3239, in launchConsoleProcess
    proc = subprocess.Popen(args, env=startupEnv, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True, cwd=cwd)
  File "/home/andinet/bin/LungAir-d/python-install/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/andinet/bin/LungAir-d/python-install/lib/python3.9/subprocess.py", line 1744, in _execute_child
    executable = os.fsencode(executable)
  File "/home/andinet/bin/LungAir-d/python-install/lib/python3.9/os.py", line 810, in fsencode
    filename = fspath(filename)  # Does type-checking of `filename`.
TypeError: expected str, bytes or os.PathLike object, not NoneType

Leengit commented:
(I typed this a few days, but never it the "comment" button)  The most recent commit implements a Qt ComboBox in the `Advanced` section of the application so that one can switch between `USE_PTH`, `USE_LOCAL_ZIP`, and `USE_DOCKER_ZIP` without changing the source code and relaunching the application.

ebrahimebrahim commented:
Notice that it's CUDA 11.0 on deepak-krs-7910
@Leengit I think you mentioned that 11.3 is needed for monai deploy
Does the error @andinet is showing look like what you'd expect from a CUDA version being incompatible with monai deploy?

ebrahimebrahim commented on Modules/Scripted/Home/HomeLib/dependency_installer.py:
```diff
@@ -56,7 +62,14 @@ def monai_pre_install():
     slicer.util._executePythonModule('light_the_torch', ['install', 'monai'])
 
 
+cudaVersion = os.popen("nvcc --version | egrep -o 'release [0-9\.]+' | egrep -o '[0-9\.]+' | sed -r -e 's/\.//g'").read().splitlines()[0]
```
Upon launching LungAIR I got

```text
/bin/sh: 1: nvcc: not found
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/python-install/lib/python3.9/imp.py", line 169, in load_source
    module = _exec(spec, sys.modules[name])
  File "<frozen importlib._bootstrap>", line 613, in _exec
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 7, in <module>
    from HomeLib import dependency_installer
  File "/home/ebrahim/LungAIR-SuperBuild-testing-issue39/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/dependency_installer.py", line 65, in <module>
    cudaVersion = os.popen("nvcc --version | egrep -o 'release [0-9\.]+' | egrep -o '[0-9\.]+' | sed -r -e 's/\.//g'").read().splitlines()[0]
IndexError: list index out of range
```

and then the Home module failed to load. It looks like nvcc is now a dependency of LungAIR? I don't think it should be, because we ought to be able to do everything without any nvidia hardware if we want to (i.e. run inference on CPU). 

In any case, we can't have the Home module crash-- the failure should be more graceful if `nvcc` does not exist.

Leengit replied:
The `nvcc`-based code did not work and has been removed.

ebrahimebrahim commented on Modules/Scripted/Home/HomeLib/dependency_installer.py:
```diff
@@ -56,7 +62,14 @@ def monai_pre_install():
+check_and_install_monai = functools.partial(
+    check_and_install_package,
+    ["monai", "skimage", "tqdm", "PIL", "ignite", "gdown", "nibabel", "monai.transforms", "monai.deploy.core", "scipy", "numpy"],
+    f"monai[skimage,tqdm,pillow,ignite,gdown,nibabel,transformers] monai-deploy-app-sdk scipy numpy --extra-index-url https://download.pytorch.org/whl/cu{cudaVersion}",
```
We shouldn't altogether remove the option to use light-the-torch. The advantage of light-the-torch is that it insulates us from having to worry about how to find exactly the right pip install command for the user's environment. And if things change then maintenance becomes a lot easier because we just have to keep up with light-the-torch changes and not worry about changes form nvidia/pytorch/pip.

Maybe we can have two side-by-side buttons for installing monai: one for the light-the-torch approach and one for the direct approach. The direct approach could even let you just type in a cuda version, so that nvcc is not needed. Just throwing ideas out-- what do you think?

These installer buttons are temporary anyway... they really ought to be replaced by an installation that takes place during the build process. I just don't know cmake well enough to achieve that at the moment :)

Leengit replied:
I think the use of `nvcc` is an experiment that failed.  I've reverted to `light-the-torch`.

Leengit commented on Modules/Scripted/Home/HomeLib/xray.py:
```diff
@@ -193,15 +193,20 @@ def delete_nodes(self):
+      from HomeLib.segmentation_model import SegmentationModel
+      self.seg_model['model'] = SegmentationModel(self.seg_model['model_path'], backend_to_use)
+    # Use the seg_model
+    self.seg_mask_tensor, model_to_image_matrix = self.seg_model['model'].run_inference(self.get_numpy_array())
```
> Maybe the real problem is that the `Xray` class is coupled to `SegmenationModel`

I too think there is good reason to look at this architecture and decide whether `Xray` should continue to have a `seg_model` attribute.  I am thinking we should keep the attribute for now, and address that with a different pull request.

Leengit commented:
There is quite a bit of versioning compatibility to worry about, so I am summarizing my findings here.  For the Python code, three environments to worry about are the build environment, the environment built for LungAIRApp in which it is running, and the environment inside the docker image.  Within these three environments, potential versioning compatibility issues arise for the version of `Python`, the version of `CUDA`, and the version of `torch`.

I find that things are working when:
1. the build and docker environments are using `Python` 3.8.10 and the runtime is using `Python` 3.9.  So there is at least a little flexibility there.
2. the build and runtime environments are using `CUDA` 11.7 and the docker is using `CUDA` 11.3.  So there is a decent amount of flexibility there.
3. all three environments are using `torch` 1.12.1.  The TorchScript `zip` format of `torch` 1.12.1 is incompatible with that of `torch` 1.10.  I didn't test 1.12.1 vs. other versions of `torch`, but seemingly there are some strict compatibility issues here.

To build the docker image I ran:
```shell
monai-deploy package \
    ~/git/KitwareMedical/lungair-desktop-application/Modules/Scripted/Home/HomeLib/deploy_app.py \
    --requirement docker-requirements.txt \
    --tag lung_air_app:latest \
    --model ~/git/KitwareMedical/lungair-desktop-application/Modules/Scripted/Home/Resources/PyTorchModels/LungSegmentation/model0018.zip
```
That gets needed versions for `Python` and the needed `model0018.zip` file that was created using `torch` 1.12.1 --- except that it doesn't get `torch` 1.12.1.  To get the last, I started up the docker image with
```shell
nvidia-docker run --rm --entrypoint /bin/bash lung_air_app:latest -c "sleep 86400"
```
and in a separate local shell I ran
```shell
PS="$(docker ps | fgrep lung_air_app:latest | cut -c 1-13)"
nvidia-docker exec -it $PS /bin/bash
pip install -U torch==1.12.1
exit
docker commit $PS lung_air_app:latest
docker stop $PS
```

Still to do: install the docker image I created in a publicly available location.  Test on a platform other than Ubuntu.

Leengit commented:
To create the `model0018.zip` file that is currently checked into this branch, I launched Python 3.8 that has torch==1.12.1 installed.  Within Python I ran:
```python
import torch
model_dict = torch.load(
    "Modules/Scripted/Home/Resources/PyTorchModels/LungSegmentation/model0018.pth",
    map_location=torch.device('cpu'))
torch.jit.script(model_dict['model']).save(
    "Modules/Scripted/Home/Resources/PyTorchModels/LungSegmentation/model0018.zip")
```

Leengit commented:
I got a lot of good advice from [@banesullivan](https://github.com/banesullivan) for pushing the created docker image (for the USE_DOCKER_ZIP option) to `ghcr.io` but the `docker push` is silently failing, seemingly because the image is 18GB in size.  I've started a discussion at https://github.com/Project-MONAI/monai-deploy-app-sdk/discussions/347 to solicit advice from others who have published a monai-deploy docker image.

N.B. if the solution is to add a `Dockerfile` to this repository then we can make use of Pull Request #43, which [@banesullivan](https://github.com/banesullivan) has started for us.

Leengit commented:
According to the feedback I got from Discussion https://github.com/Project-MONAI/monai-deploy-app-sdk/discussions/325, the size of the docker image should not be preventing the `docker push` to `ghcr.io`.  And when I tried again, I could successfully push to `ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest`.  Others attempting this may need adequate permissions for the repository or its organization.

With the most recent commit, the `lungair-desktop-application` code knows to find the docker image at this location.  The first use of this docker image on a computer will require that `lungair-desktop-application` downloads the 18.7 GiB image.  This can take 25 minutes on a wired 100 Mib/s connection; slower on wireless, faster on faster wired connections.  (The download will be faster if the computer already has the docker image `nvcr.io/nvidia/pytorch`.)  We could instead explicitly call `docker pull` at build time, or ask the user to invoke the `docker pull` command before launching the applicaiton; either way transferring those minutes from the first use of the docker image to another time.  For this application, the use of the docker image is overkill; but for applications where the docker-image approach makes sense, downloading on the first use will often not be the best approach.

Leengit commented:
I am now seeing compatibility issues between the CUDA version or GPU where the docker is built and the CUDA version or GPU where the docker is run.  Before a range of CUDA 11.3 to CUDA 11.7 seemed to be tolerated, but I will explore and report back.

```text
NVIDIA RTX A5000 with CUDA capability sm_86 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_70.
If you want to use the NVIDIA RTX A5000 GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/
...
  File "~/venv/lung/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 453, in _conv_forward
                            weight, bias, self.stride,
                            _pair(0), self.dilation, self.groups)
        return F.conv2d(input, weight, bias, self.stride,
               ~~~~~~~~ <--- HERE   
                        self.padding, self.dilation, self.groups)
RuntimeError: CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

Leengit commented:
> Before a range of CUDA 11.3 to CUDA 11.7 seemed to be tolerated

As best as I can tell, the version of `torch` built for a given version of `CUDA` will support GPUs that were supported during the heyday of the CUDA version.  Even though support for a newer GPU is generally backported to some older versions of CUDA, that support isn't making it through to the version of `torch` that was built for that version of `CUDA`.  In short, to install `torch` with a newer GPU, one must first install a newer version of `CUDA` even though, for other than `torch`, the older version of CUDA would work.

My attempts to install a newer version of CUDA (the modern 11.7 over a non-modern 11.3) within the docker image created by the MONAI Deploy App SDK have so far been unsuccessful.  I will ask the MONAI Deploy App SDK folks for their wisdom.

Leengit commented:
The response to my discussion question https://github.com/Project-MONAI/monai-deploy-app-sdk/discussions/358 is that monai-deploy is not supporting the latest CUDA and latest GPUs.  A fix was requested by the responder via Issue https://github.com/Project-MONAI/monai-deploy-app-sdk/issues/359.

This affects monai-deploy when it is used via docker.  We should not expect a docker we create to work on the latest systems, until the underlying issue has been addressed.

ebrahimebrahim reviewed (changes requested):
I gave this another test from a clean build and it worked great, except for the docker option.
I expect that I'm probably missing some needed step for the docker option, and I'm not so sure I want to even do it given the 18GB download. Since the docker functionality is mainly a demo of how something could be done, rather than being an actually practical way to do it, I suggest not trying to fix it to get it to work on my system. Instead here is what I suggest:


I updated the README on the main brach and added [a placeholder](https://github.com/KitwareMedical/lungair-desktop-application#using-the-monai-depoly-docker-image) where some more specific instructions/warnings/prereqs can be placed for the steps needed to make the docker image work. @Leengit It would be nice if you can write into that placeholder the instructions for using the docker image. Feel free to write the readme update to main branch directly, or w/e you prefer. Thank you!

Feel free to ignore this, but for the record here's the error when I naively try the docker option without paying attention to the finicky requirements:
```txt
Checking dependencies...
--> Verifying if "docker" is installed...

--> Verifying if "ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest" is available...

Checking for MAP "ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest" locally
"ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest" not found locally.

Trying to pull from registry...
Error response from daemon: Head "https://ghcr.io/v2/kitwaremedical/lungair-desktop-application/lung_air_model_deploy/manifests/latest": unauthorized
ERROR: Unable to fetch required image.
Execution Aborted
Traceback (most recent call last):
  File "/home/ebrahim/LungAIR-SuperBuild-PR-41/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 181, in onSegmentSelectedClicked
    self.logic.segmentSelected(self.backendToUse)
  File "/home/ebrahim/LungAIR-SuperBuild-PR-41/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/Home.py", line 503, in segmentSelected
    self.xray_collection.segment_selected(backend_to_use)
  File "/home/ebrahim/LungAIR-SuperBuild-PR-41/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 444, in segment_selected
    self.selected_xray().add_segmentation(backend_to_use)
  File "/home/ebrahim/LungAIR-SuperBuild-PR-41/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/xray.py", line 209, in add_segmentation
    self.seg_mask_tensor, model_to_image_matrix = self.seg_model['model'].run_inference(self.get_numpy_array())
  File "/home/ebrahim/LungAIR-SuperBuild-PR-41/Slicer-build/lib/LungAIR-4.13/qt-scripted-modules/HomeLib/segmentation_model.py", line 140, in run_inference
    slicer.util.logProcessOutput(proc)
  File "/home/ebrahim/LungAIR-SuperBuild-PR-41/Slicer-build/bin/Python/slicer/util.py", line 3276, in logProcessOutput
    raise CalledProcessError(retcode, proc.args, output=proc.stdout, stderr=proc.stderr)
subprocess.CalledProcessError: Command '['/home/ebrahim/LungAIR-SuperBuild-PR-41/python-install/bin/monai-deploy', 'run', 'ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest', '/tmp/tmp6qhxh01a', '/tmp/tmpv6a9e_ig']' returned non-zero exit status 1.
```

Leengit commented on Modules/Scripted/Home/HomeLib/dependency_installer.py:
```diff
@@ -59,4 +65,9 @@ def monai_pre_install():
-check_and_install_monai = functools.partial(check_and_install_package, ["monai", "skimage", "tqdm"], "monai[skimage,tqdm]", monai_pre_install)
+check_and_install_monai = functools.partial(
+    check_and_install_package,
+    ["monai", "skimage", "tqdm", "PIL", "ignite", "gdown", "nibabel", "monai.transforms", "monai.deploy.core"],
```
The example I started with, https://docs.monai.io/projects/monai-deploy-app-sdk/en/latest/notebooks/tutorials/02_mednist_app.html, requires `ignite` and `gdown`.  It also requires `nibabel` even though they don't mention it explicitly -- but see https://github.com/Project-MONAI/monai-deploy-app-sdk/issues/302.

With `ignite`, I do see the explicit use  in `02_mednist_app.html`, so it is possible that we don't need it because we don't use it explicitly.  With `gdown` and `nibabel`, I don't see the explicit use, seemingly telling us that it is within the `monai-deploy` implementation that these are needed and that we will need them in our case too.  But that's easy to check; I can experiment removing these dependencies and see what happens.

ebrahimebrahim replied:
Thanks for checking!

Leengit replied:
We don't need `ignite`, `gdown`, and `nibabel`.  The commit that I just pushed now omits them.

Leengit commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -2,44 +2,71 @@
 
 class SegmentationModel:
-  def __init__(self, load_path):
+  class ModelSource(enum.Enum):
```
Will do.

Leengit replied:
In the commit that I just pushed, I used
```python
        TORCHSCRIPT_ONLY = 'TorchScript only'
        LOCAL_DEPLOY = 'MONAI Deploy via command line'
        DOCKER_DEPLOY = 'MONAI Deploy via docker image'
```

Leengit commented on Modules/Scripted/Home/Home.py:
```diff
@@ -106,6 +106,15 @@ def add_install_button(package_name:str, install_function:str):
     add_install_button("pandas", dependency_installer.check_and_install_pandas)
     add_install_button("matplotlib", dependency_installer.check_and_install_matplotlib)
+    backendComboBox = qt.QComboBox()
+    backendComboBox.addItems(["USE_PTH", "USE_LOCAL_ZIP", "USE_DOCKER_ZIP"])
```
Will do.

Leengit replied:
In the commit that I just pushed, I used
```python
        TORCHSCRIPT_ONLY = 'TorchScript only'
        LOCAL_DEPLOY = 'MONAI Deploy via command line'
        DOCKER_DEPLOY = 'MONAI Deploy via docker image'
```

Leengit replied:
I updated this `Enum`; see comments elsewhere.

Leengit commented:
> I updated the README on the main brach and added [a placeholder](https://github.com/KitwareMedical/lungair-desktop-application#using-the-monai-depoly-docker-image) where some more specific instructions/warnings/prereqs can be placed for the steps needed to make the docker image work. @Leengit It would be nice if you can write into that placeholder the instructions for using the docker image. Feel free to commit the readme update to main branch directly, or w/e you prefer. 

I will add something there.

> Feel free to ignore this, but for the record here's the error when I naively try the docker option without paying attention to the finicky requirements:

This might be easy to fix.  I suspect that you didn't `docker login ghcr.io -u YourUserName` on your computer -- why would you have bothered?  :-) -- and thus `docker pull` to `ghcr.io` didn't know who you are, and thus you don't have as much permission as I do to pull down that image.  That is, I suspect that we need to declare that docker image public and then all will be good.

ebrahimebrahim commented:
> This might be easy to fix. I suspect that you didn't `docker login ghcr.io -u YourUserName` on your computer -- why would you have bothered? :-) -- and thus `docker pull` to `ghcr.io` didn't know who you are, and thus you don't have as much permission as I do to pull down that image. That is, I suspect that we need to declare that docker image public and then all will be good.

Ah, thanks for explaining. No objection to making the docker image public

Leengit commented:
With the latest commit, I rebased to the current `master` branch, to include a recent update to the `README.md` file, and addressed several comments from above.  Still to do:

1. Change the docker image visibility to public
2. Modify `README.md` to give instructions for using the docker image.
3. Try updated CUDA via `monai-deploy package -b` as suggested by https://github.com/Project-MONAI/monai-deploy-app-sdk/discussions/358.

ebrahimebrahim commented on Modules/Scripted/Home/HomeLib/segmentation_model.py:
```diff
@@ -16,24 +16,28 @@
+      return '<%s.%s>' % (self.__class__.__name__, self.name)
+
+  class ModelSource(NoValue):
+      TORCHSCRIPT_ONLY = 'TorchScript only'
```
The pth file that was saved via `torch.save` is not TorchScript, is it? I think you have to save with `torch.jit.script` and `torch.script.save` to get it as TorchScript.

The "TorchScript only" terminology may then be somewhat misleading as well.

ebrahimebrahim replied:
Looks great, thank you!

ebrahimebrahim commented:
@Leengit All looks good; shall we merge this an open a new issue to track the remaining README task? Feel free to hit "merge" if yes. Or would you prefer to leave this PR open till the TODOs that you listed are completed?

Leengit commented:
With help from [monai-deploy-app-sdk Disscussion #358](https://github.com/Project-MONAI/monai-deploy-app-sdk/discussions/358), I have been able to get MONAI Deploy within a docker image to work even with the change in GPU between the development system (older CUDA and older GPU) and the deployment system (newer CUDA and newer GPU).  The key was to use a much newer base image with the `-b` option to `monai-deploy`, _and_ to skip the step of updating Python packages within the created image.  The working package command is:

```bash
monai-deploy package \
    lungair-desktop-application/Modules/Scripted/Home/HomeLib/deploy_app.py \
    -b nvcr.io/nvidia/pytorch:22.09-py3 \
    --tag ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest \
    --model lungair-desktop-application/Modules/Scripted/Home/Resources/PyTorchModels/LungSegmentation/model0018.zip
docker push ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest
```

Permissions are still a work in progress, but once you can successfully
```bash
docker pull ghcr.io/kitwaremedical/lungair-desktop-application/lung_air_model_deploy:latest
```
then you can run `LungAIRApp` and use the docker image on your system.
